plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/NodeMarginCalculator.java

```diff
@@ -288,48 +280,50 @@ private void computeLabelBox(final ElkRectangle labelBox, final LabelAdapter<?>
         labelBox.width = label.getSize().x;
         labelBox.height = label.getSize().y;
         
+        ElkMargin desiredNodeMargin = IndividualSpacings.getIndividualOrInherited(node, CoreOptions.NODE_LABELS_MARGIN);
+
         if (port == null) {
             // The edge is connected directly to the node
             if (incomingEdge) {
                 // Assume the edge enters the node at its western side
-                labelBox.x -= labelSpacing + label.getSize().x;
+                labelBox.x -= desiredNodeMargin.left + label.getSize().x;
             } else {
                 // Assume the edge leaves the node at its eastern side
-                labelBox.x += node.getSize().x + labelSpacing;
+                labelBox.x += node.getSize().x + desiredNodeMargin.right;
             }
         } else {
             switch (port.getSide()) {
             case UNDEFINED:
             case EAST:
                 labelBox.x += port.getSize().x
-                           + labelSpacing
+                           + desiredNodeMargin.left
                            + portLabelSpace.x
-                           + labelSpacing;
+                           + desiredNodeMargin.right; // TODO cds: I feel this is wrong (same below)
```

Comment:
I think the idea here was that there's the width of the node, then a bit of space between it and the port labels, then the width of the port labels, and then another bit of space between them and the edge end labels. In fact, this might make the node label margin a little strange, perhaps. It needs to be applied between a node and its port labels, and then again between those and the end labels... Hm...

Reply:
Oh, and outside node labels as well...